File: CS50X/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// Function prototypes
bool is_valid(long long card_number);
int get_card_type(long long card_number);

int main(void)
{
    long long card_number;

    // Prompt the user for a credit card number
    do
    {
        card_number = get_long("Number: ");
    }
    while (card_number < 0);

    // Check if the credit card number is valid
    if (is_valid(card_number))
    {
        int card_type = get_card_type(card_number);

        // Print the card type
        switch (card_type)
        {
            case 1:
                printf("AMEX\n");
                break;
            case 2:
                printf("MASTERCARD\n");
                break;
            case 3:
                printf("VISA\n");
                break;
            default:
                printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

bool is_valid(long long card_number)
{
    int sum = 0;
    int digit_count = 0;

    while (card_number > 0)
    {
        int digit = card_number % 10;

        // Double every other digit, starting from the right
        if (digit_count % 2 == 1)
        {
            digit *= 2;

            // If doubling results in a two-digit number, add the digits
            if (digit > 9)
            {
                digit = digit % 10 + 1;
            }
        }

        sum += digit;
        card_number /= 10;
        digit_count++;
    }

    // Check if the sum's last digit is 0
    return (sum % 10 == 0);
}

int get_card_type(long long card_number)
{
    // Check if it's an American Express card
    if ((card_number >= 34e13 && card_number < 35e13) || (card_number >= 37e13 && card_number < 38e13))
    {
        return 1; // AMEX
    }

    // Check if it's a MasterCard
    if ((card_number >= 51e14 && card_number < 56e14))
    {
        return 2; // MASTERCARD
    }

    // Check if it's a Visa
    if (card_number >= 4e12 && card_number < 5e12)
    {
        return 3; // VISA
    }

    return 0; // Invalid
}

File: cs50-x/sentimental-credit/credit.py
import re


# Function to check if a credit card number is valid
def is_valid_credit_card(card_number):
    # Implement Luhn’s algorithm to check the validity of the card number
    # Double every second digit from right to left
    card_number = list(map(int, card_number))
    for i in range(len(card_number) - 2, -1, -2):
        card_number[i] *= 2
        if card_number[i] > 9:
            card_number[i] -= 9

    # Sum all the digits
    total = sum(card_number)

    # Check if the total sum modulo 10 equals 0
    return total % 10 == 0


# Prompt the user for a credit card number
def main():
    card_number = input("Number: ")

    # Check if the input is entirely numeric and of the correct length
    if re.match("^\d{13,16}$", card_number):
        if is_valid_credit_card(card_number):
            if card_number[0] == "4" and (
                len(card_number) == 13 or len(card_number) == 16
            ):
                print("VISA")
            elif card_number[:2] in ["34", "37"] and len(card_number) == 15:
                print("AMEX")
            elif 51 <= int(card_number[:2]) <= 55 and len(card_number) == 16:
                print("MASTERCARD")
            else:
                print("INVALID")
        else:
            print("INVALID")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
